skbio/alignment/_path.py

```diff
@@ -0,0 +1,410 @@
+        if not (np.logical_or(bits == 0, bits == 1).all()):
+            raise ValueError("Input 'bits' must contain only zeros and ones.")
+
+        ints = bits[0] + (bits[1] << 1)
+        idx = np.append(0, np.where(ints[:-1] != ints[1:])[0] + 1)
+        lens = np.append(idx[1:] - idx[:-1], ints.size - idx[-1])
+        return cls(lens, ints[idx], np.zeros(bits.shape[0], dtype=int))
+
+    def to_bits(self):
+        r"""Unpack states into an array of bits."""
+        # This should be faster than the generic solution I guess.
+        # TODO: Pending benchmark and optimization.
+        if not np.all(np.isin(self.states, [0, 1, 2])):
+            raise ValueError(
+                "For pairwise alignment, `states` must only contain "
+                "zeros, ones, or twos."
+            )
+        return np.stack([self.states & 1, self.states >> 1])
+
+    def to_cigar(self, seqs=None):
+        r"""Get a CIGAR string representing the pairwise alignment path.
+
+        Parameters
+        ----------
+        seqs : list of skbio.sequence
+            Pair of aligned sequences to generate cigar string. If provided, will
+            distinguish match (=) and mismatch (X). Otherwise, will uniformly note
+            them as (mis)match (M).
+        """
+        # If `seqs` is provided, ensure that the sequence lengths match the alignment
+        # length.
+        if seqs:
+            if len(seqs[0]) != self.shape[1] or len(seqs[1]) != self.shape[1]:
+                raise ValueError(
+                    f"At least one of sequence lengths "
+                    f"{len(seqs[0]), len(seqs[1])} does not match "
+                    f"alignment length ({self.shape[1]})."
+                )
+            # Also need to ensure that the sequences provided actually match the
+            # lengths and states of the instantiated object.
+            byte_arr = np.stack([x._bytes for x in seqs])
+            gap_chars = [ord(x) for x in seqs[0].gap_chars]
+            test_obj = PairAlignPath.from_bits(np.isin(byte_arr, gap_chars))
+            if (
+                self.lengths.shape != test_obj.lengths.shape
+                or (self.lengths != test_obj.lengths).any()
+            ):
+                raise ValueError(
+                    "Provided sequences do not match existing segment lengths."
+                )
+            elif (
+                self.states.shape != test_obj.states.shape
+                or (self.states != test_obj.states).any()
+            ):
+                raise ValueError(
+                    "Provided sequences do not match existing states. Consider the "
+                    "order of the provided sequences."
+                )
+
+        cigar = ""
+        lengths = self.lengths
+        gaps = np.squeeze(self.states)
+        codes = ["M", "I", "D", "P"]
+        if seqs is not None:
+            query = str(seqs[0])
+            ref = str(seqs[1])
+            for qchar, rchar in zip(query, ref):
```

Comment:
See above. This code assumes that the two sequences are of the same length, which may not be the case.